From 2f6da1717d90c6bb15ad390d34926249fb3e0a32 Mon Sep 17 00:00:00 2001
From: Chelsio Cudbg <chelsiocudbg@users.noreply.github.com>
Date: Wed, 25 Feb 2026 14:03:13 +0530
Subject: [PATCH 10/13] cxgb4: rework PCIe error handling

Consolidate adapter resource allocation and free to common functions.
Use the common functions from both registered PCIe error handler
callbacks and probe/remove callbacks.

[Most of the changes were already present, did very small changes only]
---
 drivers/net/ethernet/chelsio/cxgb4/cxgb4.h    |  2 +-
 .../net/ethernet/chelsio/cxgb4/cxgb4_main.c   |  8 +++++--
 .../net/ethernet/chelsio/cxgb4/cxgb4_uld.c    | 24 ++++++++++++-------
 drivers/net/ethernet/chelsio/cxgb4/sge.c      | 10 --------
 drivers/net/ethernet/chelsio/cxgb4/t4_hw.c    | 18 +++++++-------
 5 files changed, 32 insertions(+), 30 deletions(-)

diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4.h b/drivers/net/ethernet/chelsio/cxgb4/cxgb4.h
index 1946fb1c2237..b20380c5ea09 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4.h
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4.h
@@ -791,6 +791,7 @@ enum {                                 /* adapter flags */
 	CXGB4_ROOT_NO_RELAXED_ORDERING	= (1 << 10),
 	CXGB4_SHUTTING_DOWN		= (1 << 11),
 	CXGB4_SGE_DBQ_TIMER		= (1 << 12),
+	CXGB4_RES_ALLOC_DONE		= (1 << 13),
 };
 
 enum {
@@ -2340,7 +2341,6 @@ void t4_tp_tm_pio_read(struct adapter *adap, u32 *buff, u32 nregs,
 void t4_tp_mib_read(struct adapter *adap, u32 *buff, u32 nregs,
 		    u32 start_index, bool sleep_ok);
 
-void t4_uld_mem_free(struct adapter *adap);
 int t4_uld_mem_alloc(struct adapter *adap);
 void t4_uld_clean_up(struct adapter *adap);
 void t4_register_netevent_notifier(void);
diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
index 7bddc0de73c4..8028d51b309c 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
@@ -4395,7 +4395,7 @@ static struct fw_info *find_fw_info(int chip)
 /*
  * Phase 0 of initialization: contact FW, obtain config, perform basic init.
  */
-static int adap_init0(struct adapter *adap, int vpd_skip)
+static int adap_init0(struct adapter *adap, bool vpd_skip)
 {
 	unsigned int chip_ver = CHELSIO_CHIP_VERSION(adap->params.chip);
 	struct fw_caps_config_cmd caps_cmd;
@@ -5656,7 +5656,7 @@ static void free_some_resources(struct adapter *adapter)
 #endif
 	cxgb4_disable_irqs(adapter);
 
-	for_each_port(adapter, i)
+	for_each_port(adapter, i){
 		if (adapter->port[i]) {
 			struct port_info *pi = adap2pinfo(adapter, i);
 
@@ -5666,6 +5666,7 @@ static void free_some_resources(struct adapter *adapter)
 			kfree(adap2pinfo(adapter, i)->rss);
 			free_netdev(adapter->port[i]);
 		}
+	}
 	if (adapter->flags & CXGB4_FW_OK)
 		cxgb4_common_fw_free(adapter);
 }
@@ -6600,6 +6601,9 @@ void cxgb4_adap_shutdown(struct adapter *adapter)
 			if (adapter->port[i]->reg_state == NETREG_REGISTERED)
 				cxgb_close(adapter->port[i]);
 
+		if (!(adapter->flags & CXGB4_RES_ALLOC_DONE))
+			return;
+
 		rtnl_lock();
 		cxgb4_mqprio_stop_offload(adapter);
 		rtnl_unlock();
diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c
index 5c3d9cd8e1f7..2a32f75a8a95 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c
@@ -2671,6 +2671,20 @@ request_msix_queue_irqs_uld(struct adapter *adap, unsigned int uld_type)
 	return err;
 }
 
+static void
+stop_sge_rxqs_uld(struct adapter *adap, unsigned int uld_type)
+{
+	struct sge_uld_rxq_info *rxq_info = adap->sge.uld_rxq_info[uld_type];
+	unsigned int idx;
+
+	for_each_uldrxq(rxq_info, idx)
+		t4_iq_stop(adap, adap->mbox, adap->pf, 0, FW_IQ_TYPE_FL_INT_CAP,
+			   rxq_info->uldrxq[idx].rspq.cntxt_id,
+			   rxq_info->uldrxq[idx].fl.size ?
+			   rxq_info->uldrxq[idx].fl.cntxt_id : 0xffff,
+			   0xffff);
+}
+
 static void
 free_msix_queue_irqs_uld(struct adapter *adap, unsigned int uld_type)
 {
@@ -2738,19 +2752,13 @@ int t4_uld_mem_alloc(struct adapter *adap)
 	return -ENOMEM;
 }
 
-void t4_uld_mem_free(struct adapter *adap)
-{
-	struct sge *s = &adap->sge;
-
-	kfree(s->uld_txq_info);
-	kfree(s->uld_rxq_info);
-}
-
 /* This function should be called with uld_mutex taken. */
 static void cxgb4_shutdown_uld_adapter(struct adapter *adap, enum cxgb4_uld_type type)
 {
 	if (adap->uld_handle[type]) {
 
+		stop_sge_rxqs_uld(adap, type);
+
 		if (adap->flags & CXGB4_FULL_INIT_DONE)
 			quiesce_rx_uld(adap, type);
 
diff --git a/drivers/net/ethernet/chelsio/cxgb4/sge.c b/drivers/net/ethernet/chelsio/cxgb4/sge.c
index 886863cebbf1..986b1aa7fe48 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/sge.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/sge.c
@@ -5165,16 +5165,6 @@ void t4_free_sge_resources(struct adapter *adap)
 		}
 	}
 
-	if (adap->sge.fw_evtq.desc) {
-		free_rspq_fl(adap, &adap->sge.fw_evtq, NULL);
-		if (adap->sge.fwevtq_msix_idx >= 0)
-			cxgb4_free_msix_idx_in_bmap(adap,
-						    adap->sge.fwevtq_msix_idx);
-	}
-
-	if (adap->sge.nd_msix_idx >= 0)
-		cxgb4_free_msix_idx_in_bmap(adap, adap->sge.nd_msix_idx);
-
 	if (adap->sge.intrq.desc)
 		free_rspq_fl(adap, &adap->sge.intrq, NULL);
 
diff --git a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
index 3cbf47837586..3f1811a2f503 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
@@ -8177,20 +8177,20 @@ int t4_fw_bye(struct adapter *adap, unsigned int mbox)
 }
 
 /**
- *	t4_early_init - ask FW to initialize the device
- *	@adap: the adapter
- *	@mbox: mailbox to use for the FW command
+ *     t4_early_init - ask FW to initialize the device
+ *     @adap: the adapter
+ *     @mbox: mailbox to use for the FW command
  *
- *	Issues a command to FW to partially initialize the device.  This
- *	performs initialization that generally doesn't depend on user input.
+ *     Issues a command to FW to partially initialize the device.  This
+ *     performs initialization that generally doesn't depend on user input.
  */
 int t4_early_init(struct adapter *adap, unsigned int mbox)
 {
-	struct fw_initialize_cmd c;
+       struct fw_initialize_cmd c;
 
-	memset(&c, 0, sizeof(c));
-	INIT_CMD(c, INITIALIZE, WRITE);
-	return t4_wr_mbox(adap, mbox, &c, sizeof(c), NULL);
+       memset(&c, 0, sizeof(c));
+       INIT_CMD(c, INITIALIZE, WRITE);
+       return t4_wr_mbox(adap, mbox, &c, sizeof(c), NULL);
 }
 
 /**
-- 
2.39.1

